0011-chunking-o1js.md

```diff
@@ -0,0 +1,97 @@
+
+1. Adjusting OCaml bindings to Pickles - [pickles_bindings.ml](https://github.com/o1-labs/o1js-bindings/blob/1db0524c5c7c93f0df9db3fb3bdfa0414bdaf90b/ocaml/lib/pickles_bindings.ml#L563)
+
+   - Modifying the `pickles_compile` function to accept an additional parameter `num_chunks`, which will then be passed into `Pickles.compile_promise`. `num_chunks` will be calculates in o1js directly.
+
+2. Adapting the compile function in [proof_system.ts](https://github.com/o1-labs/o1js/blob/0d82ad1d67417c72b06fc8456a0a2b54bcc46652/src/lib/proof_system.ts)
+
+   - `compileProgram` function should accept a new argument `numChunks` which represents the amount of chunks needed to compile the circuit
+   - `numChunks` inside of `compileProgram` will then be passed into `Pickles.compile` via the bindings layer
+
+3. Before calling into `compileProgram`, the number of chunks must be calculated. This can be done by utilizing the `analyzeMethods` function which exists on both `SmartContract` and `ZkProgram`.
+   - For `ZkProgram`, this change can be done in `proof_systems.ts` and the `ZkProgram` function
+   - For `SmartContract`, the change is needed in [zkapp.ts](https://github.com/o1-labs/o1js/blob/0d82ad1d67417c72b06fc8456a0a2b54bcc46652/src/lib/zkapp.ts) - in the `SmartContract.compile` method.
+
+- ## Changing the type definitions of the bindings to represent the change in the `pickles_compile` function
+
+## Test plan and functional requirements
+
+1. Testing Goals and Objectives:
+
+- o1js can compile and prove circuits that are bigger than 2^16 constraints, up to ideally 2^32 (the new limit introduced by chunking)
+- A proof using a larger circuit limit can be deployed to and verified on a testnet
+- Circuits that exceed the old limit of 2^16 constraints can be recursively verified in each other
+
+2. Testing Approach:
+
+- multiple tests in o1js that test the new limits of the circuits. Tests in Kimchi and Pickles already exists, they can be used as [examples](https://github.com/MinaProtocol/mina/blob/develop/src/lib/pickles/test/chunked_circuits/test_chunked_circuits.ml)
+
+3. Testing Scope:
+
+- Proofs can be constructed and verified locally as well as deployed to a testnet.
+
+## Drawbacks
+
+Proving larger circuits will increase proving time and memory consumption. Given the memory limit in most browsers and WASM, this could cause problems for users trying to prove very large circuits.
+
+## Rationale and alternatives
+
+Not exposing Chunking and continuing to rely on recursion for large computations. This adds a significant overhead for developers and adds complexity of applications.
+
+## Prior art
+
+Chunking working in [OCaml and Pickles](https://github.com/MinaProtocol/mina/blob/develop/src/lib/pickles/test/chunked_circuits/test_chunked_circuits.ml) directly.
+
+## Unresolved questions
```

Comment:
It looks like some additional work may be required in pickles to resolve these issues. If so, we should add these to the Technical Implementation section above to account for all the work. The added implementation items can simply give a brief overview of the neceessary fix and refer to this section (Unresolved questions) for further info